lightning/src/routing/scoring.rs

```diff
@@ -678,25 +793,104 @@ impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiqui
 			} else {
 				let negative_log10_times_2048 =
 					approx::negative_log10_times_2048(numerator, denominator);
-				self.combined_penalty_msat(amount_msat, negative_log10_times_2048, params)
+				Self::combined_penalty_msat(amount_msat, negative_log10_times_2048,
+					params.liquidity_penalty_multiplier_msat,
+					params.liquidity_penalty_amount_multiplier_msat)
+			}
+		};
+
+		if params.historical_liquidity_penalty_multiplier_msat != 0 ||
+		   params.historical_liquidity_penalty_amount_multiplier_msat != 0 {
+			// If historical penalties are enabled, calculate the penalty by walking the set of
+			// historical liquidity bucket (min, max) combinations (where the min is less than the
+			// max) and, for each, calculate the probability of success given our payment amount,
+			// then total the weighted average probability of success.
+			//
+			// We use a sliding scale to decide which value will be used for a given bucket - the
+			// first bucket receives zero (ie the lower-bound of the bucket), the last bucket
+			// receives nine (ie the upper-bound of the bucket). This avoids failing to assign
+			// penalties to channels at the edges.
+			//
+			// If we used the bottom edge of buckets, we'd end up never assigning any penalty at
+			// all to such a channel when sending less than ~0.19% of the channel's capacity (e.g.
+			// ~200k sats for a 1 BTC channel!).
+			//
+			// If we used the middle of each bucket we'd never assign any penalty at all when
+			// sending less than 1/16th of a channel's capacity, or 1/8th if we used the top of the
+			// bucket.
+			let mut total_valid_points_tracked = 0;
+			let decays = self.now.duration_since(*self.last_updated).as_secs()
+				.checked_div(params.no_updates_historical_half_life.as_secs())
+				.map(|decays| cmp::min(decays, u32::max_value() as u64) as u32)
```

Comment:
I believe this can be made at least a little bit more concise as 
```suggestion
				.map_or(u32::max_value(), |decays| cmp::min(decays, u32::max_value() as u64) as u32);
```